pkg/syncer/storage/storage_pv_controller.go

```diff
@@ -0,0 +1,184 @@
+	getDownstreamPersistentVolumeClaim    func(persistentVolumeClaimName, persistentVolumeClaimNamespace string) (runtime.Object, error)
+	getUpstreamPersistentVolume           func(clusterName logicalcluster.Name, persistentVolumeName string) (runtime.Object, error)
+}
+
+// NewPersistentVolumeSyncer returns a new storage persistent volume syncer controller.
+func NewPersistentVolumeSyncer(
+	syncerLogger logr.Logger,
+	syncTargetWorkspace logicalcluster.Name,
+	syncTargetName, syncTargetKey string,
+	upstreamClient kcpdynamic.ClusterInterface,
+	downstreamClient dynamic.Interface,
+	downstreamKubeClient *kubernetes.Clientset,
+	upstreamInformers kcpdynamicinformer.DynamicSharedInformerFactory,
+	downstreamInformers dynamicinformer.DynamicSharedInformerFactory,
+	syncTargetUID types.UID,
+) (*PersistentVolumeController, error) {
+
+	c := &PersistentVolumeController{
+		queue:             workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), persistentVolumeControllerName),
+		downstreamClient:  downstreamClient,
+		upstreamInformers: upstreamInformers,
+		upstreamClient:    upstreamClient,
+		syncTarget: syncTargetSpec{
+			name:      syncTargetName,
+			workspace: syncTargetWorkspace,
+			uid:       syncTargetUID,
+			key:       syncTargetKey,
+		},
+		updateDownstreamPersistentVolumeClaim: func(ctx context.Context, persistentVolumeClaim *corev1.PersistentVolumeClaim) (*corev1.PersistentVolumeClaim, error) {
+			var err error
+			err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+				persistentVolumeClaim, err = downstreamKubeClient.CoreV1().PersistentVolumeClaims(persistentVolumeClaim.Namespace).Update(ctx, persistentVolumeClaim, metav1.UpdateOptions{})
+				return err
+			})
+			if err != nil {
+				return nil, fmt.Errorf("failed to update PersistentVolumeClaim %s/%s: %w", persistentVolumeClaim.Namespace, persistentVolumeClaim.Name, err)
+			}
+
+			return persistentVolumeClaim, nil
+		},
+		getDownstreamPersistentVolumeClaim: func(persistentVolumeClaimName, persistentVolumeClaimNamespace string) (runtime.Object, error) {
+			return downstreamInformers.ForResource(pvcSchemeGroupVersion).Lister().ByNamespace(persistentVolumeClaimNamespace).Get(persistentVolumeClaimName)
+		},
+		getUpstreamPersistentVolume: func(clusterName logicalcluster.Name, persistentVolumeName string) (runtime.Object, error) {
+			// We don't need the namespace since the PV is cluster scoped.
+			return upstreamInformers.ForResource(pvSchemeGroupVersion).Lister().ByCluster(clusterName).Get(persistentVolumeName)
+		},
+	}
+
+	logger := logging.WithReconciler(syncerLogger, persistentVolumeControllerName)
+
+	// Add the PV informers to the controller to react to upstream PV events.
+	logger.V(2).Info("Setting up informer", pvSchemeGroupVersion.String())
+	upstreamInformers.ForResource(pvSchemeGroupVersion).Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
```

Comment:
We should add a TODO here, so that we in fact point to the Upsyncer virtual workspace informer here (once @bipuladh  PR is available)